source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,393 @@
+  std::unordered_map<uint32_t, std::unordered_map<uint32_t, uint32_t>> result;
+
+  const auto* inst_block = ir_context->get_instr_block(inst);
+
+  for (uint32_t i = 0; i < inst->NumInOperands(); ++i) {
+    const auto& operand = inst->GetInOperand(i);
+    if (operand.type != SPV_OPERAND_TYPE_ID) {
+      continue;
+    }
+
+    auto* dependency = ir_context->get_def_use_mgr()->GetDef(operand.words[0]);
+    assert(dependency && "|inst| depends on invalid id");
+
+    if (ir_context->get_instr_block(dependency) == inst_block &&
+        dependency->opcode() == SpvOpPhi &&
+        !result.count(dependency->result_id())) {
+      std::unordered_map<uint32_t, uint32_t> label_id_to_result_id;
+
+      for (uint32_t j = 1; j < dependency->NumInOperands(); j += 2) {
+        label_id_to_result_id[dependency->GetSingleWordInOperand(j)] =
+            dependency->GetSingleWordInOperand(j - 1);
+      }
+
+      result[dependency->result_id()] = std::move(label_id_to_result_id);
+    }
+  }
+
+  return result;
+}
+
+}  // namespace
+
+TransformationPropagateInstructionUp::TransformationPropagateInstructionUp(
+    const protobufs::TransformationPropagateInstructionUp& message)
+    : message_(message) {}
+
+TransformationPropagateInstructionUp::TransformationPropagateInstructionUp(
+    uint32_t block_id,
+    const std::unordered_map<uint32_t, uint32_t>& predecessor_id_to_fresh_id) {
+  message_.set_block_id(block_id);
+  message_.mutable_predecessor_id_to_fresh_id()->insert(
+      predecessor_id_to_fresh_id.begin(), predecessor_id_to_fresh_id.end());
+}
+
+bool TransformationPropagateInstructionUp::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // Check that we can apply this transformation to the |block_id|.
+  if (!IsApplicableToTheBlock(ir_context, message_.block_id())) {
+    return false;
+  }
+
+  std::vector<uint32_t> maybe_fresh_ids;
+  for (auto id : ir_context->cfg()->preds(message_.block_id())) {
+    // Each predecessor must have a fresh id in the |predecessor_id_to_block_id|
+    // map.
+    if (!message_.predecessor_id_to_fresh_id().contains(id)) {
+      return false;
+    }
+
+    maybe_fresh_ids.push_back(message_.predecessor_id_to_fresh_id().at(id));
+  }
+
+  // All ids must be unique and fresh.
+  return !fuzzerutil::HasDuplicates(maybe_fresh_ids) &&
+         std::all_of(maybe_fresh_ids.begin(), maybe_fresh_ids.end(),
+                     [ir_context](uint32_t id) {
+                       return fuzzerutil::IsFreshId(ir_context, id);
+                     });
+}
+
+void TransformationPropagateInstructionUp::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* inst = GetInstructionToPropagate(ir_context, message_.block_id());
+  assert(inst &&
+         "The block must have at least one supported instruction to propagate");
+
+  // |inst| might depend on OpPhi instructions from the same basic block.
```

Comment:
I would add here: "That's OK, because by looking at the OpPhi instruction we can tell which of its operands we should use instead when we propagate the instruction into each predecessor."

Reply:
Fixed in c74f0e3.